### PR TITLE
Instance profile provider retries

### PR DESCRIPTION
**Pull request by srchase**
resolves https://github.com/aws/aws-sdk-php/issues/1664

Implements exponential backoff when invalid JSON is returned to InstanceProfileProvider.

By default, the provider will retry three times.

Retries can be adjusted on the provider with the ```retries``` config option:
```
$provider = new InstanceProfileProvider([
    'retries' => 5
]);
```
And can be disabled by setting to ```0```:
```
$provider = new InstanceProfileProvider([
    'retries' => 0
]);
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/Credentials/InstanceProfileProvider.php [hunk 1]
```diff
@@ -24,18 +25,27 @@ class InstanceProfileProvider
...
      * - profile: Optional EC2 profile name, if known.
+     * - retries: Optional number of retries to be attempted.
      *
      * @param array $config Configuration options.
      */
     public function __construct(array $config = [])
     {
         $this->timeout = isset($config['timeout']) ? $config['timeout'] : 1.0;
         $this->profile = isset($config['profile']) ? $config['profile'] : null;
+        $this->retries = isset($config['retries']) ? $config['retries'] : 0;
```
Comment (howardlopez) [inline src/Credentials/InstanceProfileProvider.php]: Should this retry by default?

Comment (srchase) [reply src/Credentials/InstanceProfileProvider.php]: Once?  Or preserve the existing behavior of this not getting retried by default.

Comment (srchase) [reply src/Credentials/InstanceProfileProvider.php]: 3 sounds like a good compromise.

File: src/Credentials/InstanceProfileProvider.php [hunk 2]
```diff
@@ -52,8 +62,22 @@ public function __invoke()
...
                 $this->profile = (yield $this->request(self::CRED_PATH));
             }
-            $json = (yield $this->request(self::CRED_PATH . $this->profile));
-            $result = $this->decodeResult($json);
+            $result = null;
+            while ($result == null) {
+                try {
+                    $json = (yield $this->request(self::CRED_PATH . $this->profile));
+                    $result = $this->decodeResult($json);
+                } catch (InvalidJsonException $e) {
```
Comment (howardlopez) [inline src/Credentials/InstanceProfileProvider.php]: Should there be a retry for things beyond invalid JSON?

Comment (srchase) [reply src/Credentials/InstanceProfileProvider.php]: I think for now, we should limit this to the JSON error.  The original issue that precipitated needing to retry was for invalid JSON.  We may want to extend this to throttling errors in the future, but will want to be careful to not exacerbate that issue.

File: tests/Credentials/InstanceProfileProviderTest.php [hunk 3]
```diff
@@ -200,4 +202,63 @@ public function testThrowsExceptionOnInvalidJsonResponse()
...
+    public function testNetworkingErrorsAreRetried()
+    {
+        $retries = 1;
+
+        $t = time() + 1000;
+        $result = json_encode($this->getCredentialArray('foo', 'baz', null, "@{$t}"));
+        $responses = [new Response(200, [], Psr7\stream_for($result))];
+
+        $client = function () use (&$retries, $responses) {
+            if (0 === --$retries) {
```
Comment (howardlopez) [inline tests/Credentials/InstanceProfileProviderTest.php]: With the `--` before `$retries`, won't this only ever execute once without ever hitting the rejection below?

Comment (srchase) [reply tests/Credentials/InstanceProfileProviderTest.php]: Yes, should be a post-decrement so the rejection is thrown on the first request.

**Pull request merged**